src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.cc

```diff
@@ -365,6 +371,85 @@ FileWatcherCertificateProvider::ReadIdentityKeyCertPairFromFiles(
   return absl::nullopt;
 }
 
+DataWatcherCertificateProvider::DataWatcherCertificateProvider(
+    std::string root_certificate,
+    grpc_core::PemKeyCertPairList pem_key_cert_pairs)
+    : StaticDataCertificateProvider(std::move(root_certificate),
+                                    std::move(pem_key_cert_pairs)) {}
+
+absl::Status DataWatcherCertificateProvider::SetRootCertificate(
+    const std::string& root_certificate) {
+  grpc_core::MutexLock lock(&mu_);
+  if (root_certificate_ == root_certificate) {
+    return absl::InvalidArgumentError("Root Certificate has not changed.");
+  }
+  root_certificate_ = root_certificate;
+  ExecCtx exec_ctx;
+  grpc_error_handle root_cert_error = GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+      "Root Certificates are watched, while their contents are empty.");
+  for (const auto& p : watcher_info_) {
+    const std::string& cert_name = p.first;
+    const WatcherInfo& info = p.second;
+    absl::optional<std::string> root_to_report;
+    if (info.root_being_watched) {
+      root_to_report = root_certificate_;
+    }
+    if (root_to_report.has_value() && !root_to_report.value().empty()) {
+      distributor_->SetKeyMaterials(cert_name, std::move(root_to_report),
+                                    absl::nullopt);
+    }
+    if (info.root_being_watched && root_certificate_.empty()) {
+      distributor_->SetErrorForCert(cert_name, GRPC_ERROR_REF(root_cert_error),
+                                    absl::nullopt);
+    }
+  }
+  GRPC_ERROR_UNREF(root_cert_error);
+  return absl::OkStatus();
+}
+
+absl::Status DataWatcherCertificateProvider::SetKeyCertificatePairs(
+    const PemKeyCertPairList& pem_key_cert_pairs) {
+  grpc_core::MutexLock lock(&mu_);
+  if (pem_key_cert_pairs == pem_key_cert_pairs_) {
+    return absl::InvalidArgumentError(
+        "The Key-Cert pair list has not changed.");
+  }
+  absl::StatusOr<bool> all_matched_or =
+      PrivateKeyAndCertificateMatch(pem_key_cert_pairs);
+  if (!all_matched_or.ok()) {
+    std::string error_message = std::string(all_matched_or.status().message());
+    gpr_log(GPR_ERROR, "Key-Cert pair list error: %s", error_message.c_str());
```

Comment:
We can avoid error_message variable here, and the code can be written as
gpr_log(GPR_ERROR, "Key-Cert pair list error: %s", std::string(all_matched_or.status().message()).c_str());

I wasn't sure, do we need these logs, aren't we anyway returning the error back to application?

Reply:
Noted, thanks.

Reply:
Also, @ZhenLian advised I log these errors to help the user with debugging.